### ISSUE TITLE
Backend unloading causes crashes

### ISSUE THREAD

**Issue by msobczak** [SOCI/soci#744]
Hi All,

I got reports of crashes related to backend unloading, in this area:

https://github.com/SOCI/soci/blob/master/src/core/backend-loader.cpp#L335

Interestingly, the crash depends on what the program did with the backend (whether the session was successfully created, etc.), which makes me think that it can relate to some resources that are kept cashed internally by the native API in anticipation of further connections, etc., and which might not be under control of the user code at all. This in turn might mean that the whole idea of backend unloading is generally unfeasible - and in fact pointless, as backends are unloaded at the end of the program (from ~static_state_mgr()), when there is no further program state that could benefit from such action. That is, even though there is an added value from the dynamic loading of backends, there is no added value from unloading them (and there is a potential risk when doing it).

So instead of spending my time debugging this issue any deeper, my idea is to remove the backend unloading functionality altogether. Let me know if you have any more elaborate thoughts about this problem.

**Comment by mloskot** [comment]
No thoughts, no objections. Just let me know if you aim it for SOCI 4, in case I managed to start packaging 4.0 any time soon.

**Comment by vadz** [comment]
I don't really understand how could any resource remain when unloading the backend should also unload the database-specific DLL owning these resources.

I think unloading backends may be somewhat useful for long-running programs capable of connecting to different kinds of databases, so it seems a pity to remove this functionality entirely without even understanding why does it need to be done. OTOH I'm perfectly fine with not unloading them on shutdown, this indeed doesn't seem to have any advantages.

**Comment by msobczak** [comment]
It is possible for the native API library (be it OCI or libpq or whatever) to retain whatever resources they like in anticipation of future reuse. For example, creating and closing a session might presumably create a new network connection, but there is no provision for that network connection to be closed (and memory cleaned, etc.) when the session is closed and the library might want to keep the connection open for the next session and in fact the library might want to keep it alive just in case there is a new session created in the future. We have no control over these kind of optimizations and I can imagine them messing with the process when the library is unloaded.

In general, I disagree with the usefulness of the dynamic loading in SOCI anyway. This mechanism makes sense when host and plugins come from different sources (vendors). This is the case for example with web browser plugins. But the whole of SOCI comes from the same repository and there is no benefit for the user to have backends separately loaded. Their size is also small enough to make any savings irrelevant - that is, there is no benefit for the application to unload the backend if its size is multiple orders of magnitudes smaller than the available memory. This is one of those parts of the library that I consider to be overengineered. And I'm happy to streamline it when it causes troubles.

**Comment by vadz** [comment]
> It is possible for the native API library (be it OCI or libpq or whatever) to retain whatever resources they like in anticipation of future reuse. 

Yes, but when the backend shared library is unloaded, `liboci.so` or `libpq.so` or whatever is (or should be) unloaded as well. And it really, really shouldn't crash on unload. So this simply doesn't seem to explain the reason for the crash at all to me.

> In general, I disagree with the usefulness of the dynamic loading in SOCI anyway.

I disagree but this is really not the salient point anyhow. I don't feel good about having a mysterious crash in the library.
